typescript_test/test.ts

```diff
@@ -0,0 +1,376 @@
+import {
+  createSelector,
+  defaultMemoize,
+  createSelectorCreator,
+  createStructuredSelector,
+  ParametricSelector,
+} from '../src/index';
+
+
+
+function testSelector() {
+  type State = {foo: string};
+
+  const selector = createSelector(
+    (state: State) => state.foo,
+    (foo) => foo,
+  );
+
+  const foo: string = selector({foo: 'bar'});
+
+  // typings:expect-error
+  selector({foo: 'bar'}, {prop: 'value'});
+
+  // typings:expect-error
+  const num: number = selector({foo: 'bar'});
+
+  // typings:expect-error
+  createSelector(
+    (state: {foo: string}) => state.foo,
+    (state: {bar: number}) => state.bar,
+    (foo, bar) => 1
+  );
+}
+
+type Component<P> = (props: P) => any;
+
+declare function connect<S, P, R>(selector: ParametricSelector<S, P, R>):
+  (component: Component<P & R>) => Component<P>;
+
+function testConnect() {
+  connect(
+    createSelector(
+      (state: {foo: string}) => state.foo,
+      foo => ({foo}),
+    )
+  )(props => {
+    // typings:expect-error
+    props.bar;
+
+    const foo: string = props.foo;
+  });
+
+  const connected = connect(
+    createSelector(
+      (state: {foo: string}) => state.foo,
+      (state: never, props: {bar: number}) => props.bar,
+      (foo, bar) => ({foo, baz: bar}),
+    )
+  )(props => {
+    const foo: string = props.foo;
+    const bar: number = props.bar;
+    const baz: number = props.baz;
+    // typings:expect-error
+    props.fizz;
+  });
+
+  connected({bar: 42});
+
+  // typings:expect-error
+  connected({bar: 42, baz: 123});
+}
+
+function testInvalidTypeInCombinator() {
+  // typings:expect-error
+  createSelector(
+    (state: {foo: string}) => state.foo,
+    (foo: number) => foo,
+  );
+
+  // typings:expect-error
+  createSelector(
+    (state: {foo: string, bar: number, baz: boolean}) => state.foo,
+    state => state.bar,
+    state => state.baz,
+    (foo: string, bar: number, baz: boolean, fizz: string) => {}
+  );
+}
+
+function testRestParameters() {
+  type MyState = {foo: string, bar: number};
+  createSelector(
+    (state: MyState) => state.foo,
+    (state: MyState) => state.foo,
+    (state: MyState) => state.foo,
+    (state: MyState) => state.foo,
+    (state: MyState) => state.foo,
+    (state: MyState) => state.bar,
+    (state: MyState) => state.bar,
+    (state: MyState) => state.bar,
+    (state: MyState) => state.bar,
+    (state: MyState) => state.bar,
+    (foo1: string, foo2: string, foo3: string, foo4: string, foo5: string,
+     bar1: number, bar2: number, bar3: number, bar4: number, bar5: string) => {
```

Comment:
bar5 should be number here? But as discussed the type inference is not working with the rest arguments (this is the issue I refer to in my main comment)